java-support/src/main/java/io/cloudstate/javasupport/crdt/CrdtEntity.java

```diff
@@ -40,4 +40,12 @@
 @CloudStateAnnotation
 @Target(ElementType.TYPE)
 @Retention(RetentionPolicy.RUNTIME)
-public @interface CrdtEntity {}
+public @interface CrdtEntity {
+
+  /**
+   * Specifies the idle time of the entity before it is passivated. Zero and any negative value mean
+   * use default from configuration file. Any positive value means the entity is passivated after
+   * the idle time.
+   */
+  int passivationTimeout() default 0;
```

Comment:
Hello Guy I didn't like this option for me it should be done at the Entity registration stage and not at the entity itself.

Reply:
Hi Adriano, it is also fine for me and it is easier. I will do it at the registration stage.

Reply:
Great! Thanks

Reply:
why do you like this option?

Reply:
Why would 0 not be a valid value? And therefore passivation disabled at all?

Reply:
Doing it at the registration stage means breaking the existing registration API and all stay the same with annotations. Then we are forcing user to think about the passivation timeout for the registration. I think the option with annotations is without breaking and easier for the user. @sleipnir I would be glad to hear your idea.

Reply:
> why do you like this option?

I prefer the registration option because I understand a registration step as a configuration step for the execution service of the entity and not a property of the entity. I also prefer that the entity annotations are as simple as possible, without many options because it seems cleaner. See for example how I make the settings in Kotlin support, the user informs what he wants in the registration step and the entity remains without many changes.

```
fun main() {
    cloudstate {
        eventsourced {
            entityService = ShoppingCartEntity::class
            descriptor = Shoppingcart.getDescriptor().findServiceByName("ShoppingCart")
            additionalDescriptors = mutableListOf(Shoppingcart.getDescriptor(), Domain.getDescriptor() )
           
            // Settings that I want to inform the proxy are here:
            snapshotEvery = 1
            persistenceId = "shopping-cart"
        }

    }.start()
            .toCompletableFuture()
            .get()
}
```

Reply:
@marcellanz 0 or negative means use the default config for the passivation.

Reply:
@ralphlaude yeah saw that documented, but what is the motivation to allow a low time of 1sec, but not 0sec? <0 would be enough to piggyback to take the default config, alas setting to 0 allows the entity to disable or set it to 0.

Reply:
> I prefer the registration option because I understand a registration step as a configuration step for the execution service of the entity and not a property of the entity. I also prefer that the entity annotations are as simple as possible, without many options because it seems cleaner. See for example how I make the settings in Kotlin support, the user informs what he wants in the registration step and the entity remains without many changes.

@sleipnir it is just about Java-support here and other languages can use their own ways to do that. I have not idea about Kotlin but I don't see another option in Java. How would you configure the passivation timeout with Kotlin?

Reply:
> Doing it at the registration stage means breaking the existing registration API and all stay the same with annotations. Then we are forcing user to think about the passivation timeout for the registration. I think the option with annotations is without breaking and easier for the user. @sleipnir I would be glad to hear your idea.

I do not agree. I don't see how adding an option to the registry breaks the current API? Do the values not have defaults? Am I required to set the passivation time? If any of the answers to these questions are true then it would break the registration API, otherwise it would not.

Reply:
@ralphlaude If the user for any reason wants to use the passivation option then he is already thinking about Passivation, he is not being forced to do anything whether it be done in the annotation or in the registry phase.

Reply:
I should have written instead of
> setting to 0 allows the entity to disable or set it to 0

"setting to 0 allows the entity passivate immediately."

Reply:
> > I prefer the registration option because I understand a registration step as a configuration step for the execution service of the entity and not a property of the entity. I also prefer that the entity annotations are as simple as possible, without many options because it seems cleaner. See for example how I make the settings in Kotlin support, the user informs what he wants in the registration step and the entity remains without many changes.
> 
> @sleipnir it is just about Java-support here and other languages can use their own ways to do that. I have not idea about Kotlin but I don't see another option in Java. How would you configure the passivation timeout with Kotlin?

I know that we are talking about java support, I just gave you an example to explain the concept of what I was telling you. If you don't think it is necessary to change then ok, I just presented my arguments.

Reply:
@marcellanz the entity should not passivate immediately and it should be allowed. Setting to 0 allows the entity to passivate after 30 seconds. The`passivateTimeout` in the annotation should define the minimal value to accept.

Reply:
> @marcellanz the entity should not passivate immediately and it should be allowed. Setting to 0 allows the entity to passivate after 30 seconds. The`passivateTimeout` in the annotation should define the minimal value to accept.

This seems counter intuitive to me and should be well documented if it is assumed. Setting a property to 0 and assuming a value of 30 is somewhat unusual. If this is really the case, we cannot forget to be very clear about this in the documentation.

Reply:
@sleipnir passivation is always enabled and is implicit to the user. The `registerXXXEntity` does not  force the user to think about passivation. The new `registerXXXEntity(..., passivationTimeout)` forces the user to provide passivation timeout or at least to think about it.
For now i see two options here:
1- Annotations
2- Provide overloaded registration methods with passivationTimeout

Reply:
> @sleipnir passivation is always enabled and is implicit to the user. The `registerXXXEntity` does not force the user to think about passivation. The new `registerXXXEntity(..., passivationTimeout)` forces the user to provide passivation timeout or at least to think about it.
> For now i see two options here:
> 1- Annotations
> 2- Provide overloaded registration methods with passivationTimeout

Guy Registry phase doesn't force anyone, I already told you, sorry your argument doesn't convince me, I'm sorry. As I told you before, if you don't feel comfortable don't change, I made my contribution to the topic and I will not argue anymore.

Reply:
> I know that we are talking about java support, I just gave you an example to explain the concept of what I was telling you. If you don't think it is necessary to change then ok, I just presented my arguments.

understood and i saw how it works in Kotlin

Reply:
> This seems counter intuitive to me and should be well documented if it is assumed. Setting a property to 0 and assuming a value of 30 is somewhat unusual. If this is really the case, we cannot forget to be very clear about this in the documentation.

@sleipnir I agree with default values in annotations and it already works that way with attributes (persistentId and snapshotEvery) in `@XXXEntity`. It is also well documented. May be we should change that and explicitly work with default values in annotations.

Reply:
I agree generally about keeping the entity implementations simpler by not having too many options on the annotations. Also understand that this is the current approach in Java support, and annotation settings allow for defaults. So I'd be inclined to have both annotation and registration options right now. The registration method for annotated entity classes doesn't currently have extra options, but it makes sense to add them now. Not sure how many more options there will be in the future, but I'd consider having a settings object (so something like `EventSourcedOptions` or similar), instead of supporting multiple permutations of parameters — so the method can just be overloaded with and without the options, and so that the options object can have default or unset options.

Regarding default of 0, I'd probably interpret a timeout of 0 as disabling passivation altogether. Not sure if we would want to support disabling passivation, where the user function would control lifecycle itself or something, but I think I'd avoid 0 as being used for an unset option for selecting the proxy default.

I think it makes sense to improve the APIs and protocol approaches whenever possible. For a timeout setting, it would be good to support any kind of duration (not just int of seconds). So I would consider accepting `java.time.Duration` in the java support APIs, or `TimeUnit` variations. Annotation values are limited, can't use arbitrary classes, so one option instead of an int is to use a String and use the typesafe Config duration parsing or `Duration.parse` (which is a bit different). And can then use empty String as having the timeout unset.

For the settings that we want to have fall back to proxy-side defaults, I would support this directly in the settings — so the entity options object knows when a timeout has been set or not, and same in the protocol — and it only sends a timeout if it's been configured in the user function, rather than relying on negative numbers. And the protocol could have a duration message, with time units, to make things complete.

So I would aim for something along the lines of:

```java
// no passivation timeout set, nothing sent with entity spec to proxy
@EventSourcedEntity(persistenceId = "something")
```

```java
// passivation timeout parsed from string, sent as duration message in protocol
@EventSourcedEntity(persistenceId = "something", passivationTimeout = "10s")
```

```java
// no extra options specified, use annotation options or defaults
.registerEventSourcedEntity(
  SomeEntity.class,
  SomeProtocol.getDescriptor().findServiceByName("SomeService"),
  SomeProtocol.getDescriptor())
```

```java
// options specified, which will override annotation options or defaults
.registerEventSourcedEntity(
  SomeEntity.class,
  SomeProtocol.getDescriptor().findServiceByName("SomeService"),
  SomeProtocol.getDescriptor(),
  EventSourcedOptions.defaults().setPassivationTimeout(Duration.ofSeconds(10))
```

Reply:
I agree with the whole proposal here

Reply:
After the discussion last week, i agree fully with the proposal.

Reply:
>and it only sends a timeout if it's been configured in the user function, rather than relying on negative numbers. And the protocol could have a duration message, with time units, to make things complete.

I agree to change the protocol now and a soon as possible for such changes. I also agree on not depending on magic values.

Reply:
I think I'd like to have options for expression the following things - perhaps not immediately, but we should choose a mechanism and protocol that allows us to add support for them in future if needed:

* A specific passivation timeout, with at least millisecond resolution.
* Default - note that I think it's fine for the default to come from the support library, I don't see any reason why the proxy needs to be in control of the default.
* No passivation timeout - entities should never passivate unless rebalanced.
* Immediate passivation, this effectively means the state is reloaded for each command.
* Some way to indicate the proxy should apply smarter, more dynamic passivation strategies, for example strategies that respond to resource constraints, or perhaps auto-tune themselves based on cardinality and/or hot spot metrics.

So, basically, I think it's worth now choosing a strategy in the APIs and protocol that will allow us to add some of the features above without breaking the protocol or APIs.

Reply:
Good idea to future-proof the protocol and APIs now for other passivation options.

Reply:
Really good idea for passivation options.
More generally we should have a strategy for adding other future options (not only passivation) without breaking the protocol or the APIs.

Reply:
After some more discussion:

- passivation timeout in protocol can just be milliseconds, we don't expect more fine-grained than that
- language supports can use duration APIs or similar as appropriate, convert to milliseconds
- protocol should support expanding to other passivation strategies, where initially it could just be `timeout`, ideally we can add `never` or `immediate` or other strategies in a protocol-compatible way
- protocol should support strategies explicitly, such as `timeout`, `never` or `immediate`, and not use magic values